Functions for extracting plan id information from json query plans

mlyublena opened the pull request:
Adding the following three functions for manipulating json query plans:

```
json_presto_query_plan_ids(json jsonPlan)
json_presto_query_plan_node_children(json jsonPlan, varchar planId)
json_presto_query_plan_scrub(json jsonPlan)
```

These fill be helpful when searching for patterns in collections of query plans.
json_plan_scrub can also be used to remove noise when diffing query plans and help when investigating regressions.

## Description
<!---Describe your changes in detail-->

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Introduces new functions for manipulating Presto json query plans: json_presto_query_plan_ids, json_presto_query_plan_node_children, json_presto_query_plan_scrub
```

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

feilong-liu commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonPlanFunctions.java:
```diff
@@ -0,0 +1,253 @@
+    public static Block jsonPlanIds(@TypeParameter("ARRAY<VARCHAR>") ArrayType arrayType,
+            @SqlType(StandardTypes.JSON) Slice jsonPlan)
+    {
+        final List<JsonRenderedNode> planFragments = parseJsonPlanFragmentsAsList(jsonPlan);
```
Curious why do we have a "final" here?

mlyublena replied:
fixed

feilong-liu commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonPlanFunctions.java:
```diff
@@ -0,0 +1,253 @@
+        return planMap;
+    }
+
+    private static JsonRenderedNode parseJsonPlan(Slice jsonPlan)
```
This function is not used?

mlyublena replied:
removed

feilong-liu commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonPlanFunctions.java:
```diff
@@ -0,0 +1,253 @@
+        return planFragments;
+    }
+
+    private static JsonCodec<Map<PlanFragmentId, Map<String, JsonRenderer.JsonRenderedNode>>> constructJsonPlanMapCodec()
```
Is this using the same API used in plan serialization for logging? It will be better for consistency since I assume the input to these functions are plan json from logging.

mlyublena replied:
it uses the same basic class (JsonRenderedNode) but in the case of explain and logging the serialization happens directly from the plan node representation (to which we don't have access here)

feilong-liu commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/JsonPlanFunctions.java:
```diff
@@ -0,0 +1,253 @@
+        return constructSqlArray(arrayType, planChildren);
+    }
+
+    @Description("Scrub the query plan")
```
What do you mean by scrub here?

mlyublena replied:
added more details

mlyublena commented:
> Lets put these in a separate folder like "presto-i18n-functions".
> 
> We can name the functions so its more explicit that they are for presto query plans. Like: "presto_plan_ids"

Discussed offline: there are dependencies on presto sql plan printer and serializer which makes these hard to move to a new module. Moved to a separate directory under operator/scalar for now and renamed appropriately.
Please take another look @pranjalssh
